### DISCUSSION TITLE
[Lesson 9 - lottery] fork test error with expectRevert of testPerformUpkeepRevertsIfCheckUpkeepIsFalse

### DISCUSSION THREAD

**Discussion by LareineHan** [Cyfrin/foundry-full-course-cu#939] [Q&A] [answered]
Hi, I am getting an error with `forge test --fork-url $SEPOLIA_RPC_URL`. 
With `forge test` everything passes. 

Here is the screenshot of my error detail. 
<img width="660" alt="Screenshot 2023-09-28 at 9 59 39 AM" src="https://github.com/Cyfrin/foundry-full-course-f23/assets/128317012/48972cbd-fd71-4de9-9918-7e1392111d26">

I tried to fix it alone but nothing worked. :( So asking helps here. 
Here is the test function. 
```solidity
    function testPerformUpkeepRevertsIfCheckUpkeepIsFalse() public {
        // Arrange
        uint256 currentBalance = 0;
        uint256 numPlayers = 0;
        uint256 raffleState = 0;

        vm.warp(block.timestamp + interval + 1);
        vm.roll(block.number + 1);

        // Act & Assert ::
        vm.expectRevert( // revert means something is wrong.
            abi.encodeWithSelector(
                Raffle.Raffle__UpkeepNotNeeded.selector,
                currentBalance,
                numPlayers,
                raffleState
            )
        );
        raffle.performUpkeep("");
    }
```

And my [repo](https://github.com/LareineHan/foundry-smart-contract-lottery-f23.git) is here if needed. Thank you!!

**Comment by 0xusmanf** [comment]
Your test is failing because you are expecting `raffle.performUpkeep("");` to revert with `Raffle.Raffle__UpkeepNotNeeded(0,0,0)` whereas it is reverting with `Raffle.Raffle__UpkeepNotNeeded(0.04 ether,0,0)` while testing on Sepolia. This is happening because Raffle.sol is deploying at an address `0xA8452Ec99ce0C64f20701dB7dD3abDb607c00496` which already exists on Sepolia and has a 0.04 ether balance. The strange part is this is not even a contract address.

Why is this happening? I have no idea. I am going to look into this.

Meanwhile, can you guys explain why this is happening if you have any idea? @theirrationalone @shikhar229169 @Neftyr

Edit: Meanwhile, can you guys explain why the Raffle contract is deploying to the same address on Sepolia and how a non-contract address has anything to do with this?

Edit 2: So this is happening because in your `DeployRaffle.s.sol` the `Raffle` contract is being deployed by the `DeployRaffle` contract account and in the background the `CREATE` opcode is being used which uses `keccak256(sender, nonce);` to calculate the contract address which results in `0xA8452Ec99ce0C64f20701dB7dD3abDb607c00496` because the sender and nonce stay the same hence the same address for everybody who uses this logic.

To overcome this issue pass your deployer key to `vm.startBroadcast(deployerKey);` in your `DeployRaffle.s.sol` to generate a different address based on `deployerKey` instead of `DeployRaffle` contract account. @LareineHan

**Reply by shikhar229169** [reply]
Suppose that we are deploying our contract Y from contract X via some salt, so before even deploying contract Y we can calculate it's address beforehand. Now if someone sends eth to that address at which contract Y will be deployed (before its deployment) it doesn't mean it's an EOA. Still the contract will get deployed on that calculated address.

I performed the same forked test on sepolia and Raffle was deployed on 0x441391B18243756e57A95BFAE3B73Fc64d43C837 and I saw on explorer it has 0 eth. So, I sent 0.01 eth to it and again ran the tests and this time the test failed.

**Reply by theirrationalone** [reply]
@shikhar229169 !, Salt might be: `(Deployer's wallet address(EOA) or deployer's address(contract)) + transaction's nonce + contract's bytecode` . Or it is something unknown??

**Reply by 0xusmanf** [reply]
> Suppose that we are deploying our contract Y from contract X via some salt, so before even deploying contract Y we can calculate it's address beforehand. Now if someone sends eth to that address at which contract Y will be deployed (before its deployment) it doesn't mean it's an EOA. Still the contract will get deployed on that calculated address.

Ah I see, Thanks for explaining. Now I get why the address was showing as an EOA.

> @shikhar229169 !, Salt might be: (Deployer's wallet address(EOA) or deployer's address(contract)) + transaction's nonce + contract's bytecode . Or it is something unknown??

It uses `keccak256(sender, nonce);` as it uses `CREATE` opcode.
There is another opcode that @shikhar229169 referenced in his example `CREATE2` that uses `keccak256(0xFF, sender, salt, bytecode);`

**Comment by theirrationalone** [comment]
@usmanfarooq91 ! Thanks for the call...

`0xA8452Ec99ce0C64f20701dB7dD3abDb607c00496` might be the `RaffleTest` Address. As we know foundry works in a same pattern for all of us. So `0xA8452Ec99ce0C64f20701dB7dD3abDb607c00496` could be the `RaffleTest` Address deployed by someone else like us or by @LareineHan !.

But the issue is invalid data into revert right. Let me also know you @usmanfarooq91 ! that...
@LareineHan ! has some omissions into her `Raffle contract` that's why @LareineHan !, You're getting  issues on SEPOLIA testnet.

Issues...

1. `AutomationCompatibleInterface` import is missing. (i also mentioned it on your previous question).
2. override keyword missing from `checkUpkeep` function.
3. override keyword missing from `performUpkeep` function.

When we don't use `override` keyword, Our function looses some prior work todo. Means for instance, if we take checkupkeep function then...
`checkUpkeep` function from chainlink will appear completely different (highly decoupled) from our `checkUpkeep` function.

Now...
@LareineHan !, You should use `vm.prank` cheatcode to get that test passed on SEPOLIA.

**Reply by LareineHan** [reply]
Hi and thanks to @theirrationalone @usmanfarooq91 @shikhar229169 ! My code is still missing `AutomationCompatibleInterface` and `override ` because, by the time I am doing this test while following Patrick's lesson, he hasn't used those yet(that's what I believe). I paused and played the video very frequently to follow what he was doing, and I thought that I had caught up, but the test result was different, so I stopped watching the video and asked for help what I was missing. :( Thanks for the support and bearing with me, y'all!! Ah,, everything is so confusing.

**Comment by alfheimrShiven** [comment, marked as answer]
Hey @LareineHan you're not deploying your `Raffle.sol` contract using your own `deployerKey` i.e. your Sepolia Private Key.
Please change [this line](https://github.com/LareineHan/foundry-smart-contract-lottery-f23/blob/main/script/DeployRaffle.s.sol#L42) from 
`vm.startBroadcast(); ❌` to
`vm.startBroadcast(deployerKey); ✅`

`deployerKey` should be used to deploy `Raffle.sol` as it will deploy the Raffle on a fresh address with the starting balance as 0.
`deployerKey` is also used to create VRF subscription, fund it, and finally add `Raffle.sol` contract (that it deployed) as the consumer of VRF's funded subscription.
This should make it work as it worked for me.

@usmanfarooq91 This is probably the reason why the `Raffle.sol` contract is being deployed on an already existing Sepolia address `0xA8452Ec99ce0C64f20701dB7dD3abDb607c00496` since it was been deployed by the `DeployRaffle.s.sol` contract address and someone would have probably done the same before (not sure though).

**Reply by LareineHan** [reply]
Awesome.. finally!! how smart you are!! I changed on `DeployRaffle.s.sol` 

`vm.startBroadcast();` ❌ to
`vm.startBroadcast(deployerKey);` ✅
<img width="897" alt="Screenshot 2023-09-29 at 11 51 51 AM" src="https://github.com/Cyfrin/foundry-full-course-f23/assets/128317012/4007dabc-e0da-4ea2-951c-af64ca862ea3">

and it works! Still so confused on everything and I can't fully understand what youguys are saying 😅, 
however, phew~ now I can move on. Thank you all! 👍

**Reply by alfheimrShiven** [reply]
Hey @LareineHan I'm glad I could help. 🙌🏾

What @usmanfarooq91 was pointing out was that there is an already-existing address on [Etherscan](https://sepolia.etherscan.io/address/0xA8452Ec99ce0C64f20701dB7dD3abDb607c00496) which was exactly same as the address of your Raffle.sol contract when you were deploying without using the `deployerKey` and we were wondering how does your deployment is creating an existing address. 

Anyway, keep going. All the best!

**Reply by GabrielFilipedeOliveira** [reply]
Thanks a Lot!
